R/dominantTaxa.R

```diff
@@ -100,9 +106,56 @@ setMethod("perSampleDominantFeatures", signature = c(x = "SummarizedExperiment")
         idx <- as.list(apply(t(mat) == colMaxs(mat),1L,which))
         # Get rownames based on indices
         taxa <- rownames(mat)[unlist(idx)]
+        
         # If multiple dominant taxa were found, names contain taxa in addition to 
         # sample name. Names are converted so that they include only sample names.
         names(taxa) <- rep( names(idx), times = lengths(idx) )
+        
+        # If individual sample contains multiple dominant taxa (they have equal counts) and if 
+        # complete is FALSE, the an arbitrarily chosen dominant taxa is returned
+        if( length(taxa) > nrow(colData(x))){
+            # Store order
+            order <- unique(names(taxa))
+            # there are multiple dominant taxa in one sample (counts are equal), length
+            # of dominant is greater than rows in colData.
+            taxa <- split(taxa, rep(names(taxa), lengths(taxa)) )
+            # Order the data
+            taxa <- taxa[order]
+            
+            # If complete is set FALSE, and there are multiple dominant taxa, one of them is arbitrarily chosen
+            if( complete == FALSE) {
+                # Randomly choose one element from each sublist
+                taxa <- lapply(taxa, function(item) {
+                                if (length(item) > 1) {
+                                # If it's a list with more than one element, choose one randomly
+                                    return(sample(item, 1))
+                                } else {
+                                    return(item)
+                                }
+                            })
+                taxa <- unname(sapply(taxa, function (x) {unlist(x)}))
+                warning("Multiple dominant taxa were found for some samples. Run perSampleDominantFeatures(x, complete = TRUE) for details.")
+            }
+        }     
+        
+        # Name "Other" the features that are not included in n the most abundant in the data
+        if(!is.null(n)) {
+            flat_taxa <- unlist(taxa, recursive = TRUE)
+            top <- microbiome::top(flat_taxa, n=n)
+            # Group the rest into the "Other" category
+            
+            # complete parameter defines the output format
+            if (complete) { # return a list
```

Comment:
It can be a vector even though complete == TRUE (if there is no multiple dominant taxa)

This might work for vector and list
```

flat_taxa <- unlist(taxa, recursive = TRUE)
top <- microbiome::top(flat_taxa, n=n)
top <- names(top)

taxa <- lapply(taxa, function(x){
    ind <- sapply(x, function(y) y %in% top)
    if( any(ind) ){
        res <- x[ ind ]
    } else{
        res <- "other"
    }
    return(res)
})

if( all(lengths(taxa) == 1 ) ){
    taxa <- unlist(taxa)
}
```